core/objects.go

```diff
@@ -254,6 +274,10 @@ type Challenge struct {
 
 	// Used by dns and dvsni challenges
 	Validation *jose.JsonWebSignature `json:"validation,omitempty"`
+
+	// Contains information about URLs used or redirected to and IPs resolved and
+	// used
+	ValidationRecord *ValidationRecord `json:"validationRecord,omitempty"`
```

Comment:
I think the fact that you had to resort to `{stuff}` to indicate the contents is a great explanation of why your proposed design is underspecified. @martinthomson: It's not clear what you mean by a "generic processor." In the sense that all JSON parsing is generic, either form can be processed by a generic processor.

As I said earlier, this is bad from a langsec perspective, because validating input is not context-free: The fields that are valid in a `validationRecord` depend on the value of the neighboring `type` field, which may not even have been seen yet. For instance, this looks way less appealing:

```
 { "validationRecord": [{
    "addressesUsed": "1.2.3.4",
    "url": "example.com/foo"
  }, {
    "addressesUsed": "1.2.3.4",
    "url": "example.net/foobar"
  }
], "type": "dvsni" }
```
This is an invalid record: DVSNI doesn't support redirects. However, it's impossible to reject it as invalid until you've parsed the whole thing and get to the `"type": "dvsni"`.

@bifurcation, I do agree that we've already made this mistake with challenge fields depending on challenge types. My hope was that we could avoid making the same mistake again here, but it sounds like there's no interest. I think this change is important to get in, so I'm going to stand aside, but I still think this is the wrong choice.